[FIX] hr_holidays: prevent in-use leave type allocation requirement edit

Yoann-bary opened the pull request:
Steps to reproduce:
- Time off > Configuration > Time off types > New
- Set 'Requires Allocation' to 'No Limit'
- Takes leaves of this type
- Set 'Requires Allocation' to 'Yes'
- Approvals > Allocations > New > Select your time off type
- Edit validity period to start after the leaves you took
- Confirm > Click on the newly created allocation
- Displays -X remaining out of Y

The allocations try to cover leaves taken before the requirement was changed, when we'd expect leaves taken under No Limit mode not to require any. This is problematic because it can happen with past leaves as well, which we cannot get rid of on production databases.

Since Odoo does not track which allocation requirement mode leave are taken under nor the date and content of edits to that parameter, we cannot tell apart leaves of the same type when distributing allocated days. Tracking either would certainly require edits to the model which rules out stable versions, so instead the edits to allocation requirement should be restricted in favor of creating new time off types when we want to switch that parameter.

opw-3919886

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/180566.png)](https://mergebot.odoo.com/odoo/odoo/pull/180566)

nle-odoo reviewed (approved):
seems ok to me if it should not be changed after use

usually I think we rather use "@api.constraints", rather than an onchange, unless we still want to be able to change the value through code.

Yoann-bary commented:
| usually I think we rather use "@api.constraints", rather than an onchange

Yup that's a good point :+1: 

| seems ok to me if it should not be changed after use

I heard through BRAS that the team in charge was already planning to enforce that restriction in the future so it should be alright

Bertrand2 commented on addons/hr_holidays/models/hr_leave_type.py:
```diff
@@ -181,6 +182,18 @@ def _compute_valid(self):
 
+    @api.constrains('requires_allocation')
+    def check_allocation_requirement_edit_validity(self):
+        """ Changing the allocation requirement on a leave type which is already in use is forbidden because
+        Odoo tracks every leave of that type, even those taken prior to switching to 'Allocation Required', as
+        needing allocation which messes with the remaining days to allocate.
+
+        Raises:
+            UserError: If leaves of the given type are found
+        """
```
The message raised is sufficiently explicit, no need for that comment

Yoann-bary replied:
:+1:

Bertrand2 commented on addons/hr_holidays/models/hr_leave_type.py:
```diff
@@ -181,6 +182,18 @@ def _compute_valid(self):
+        Raises:
+            UserError: If leaves of the given type are found
+        """
+        if self.env['hr.leave'].search([('holiday_status_id', 'in', self.ids)]):
```
can be optimized:
```suggestion
        if self.env['hr.leave'].search_count([('holiday_status_id', 'in', self.ids)], limit=1):
```

Yoann-bary replied:
:+1:

Bertrand2 reviewed (changes requested):
Thanks for the PR! :smile: 
Logic seems good, I just have 2 comments on the code.
Additionally:
- don't forget to update the `.pot` file
- could you add a test ensuring the constraint works as intended?

Thanks!

Bertrand2 commented:
> I heard through BRAS that the team in charge was already planning to enforce that restriction in the future so it should be alright

I confirm! :) I was planning to do some restrictions on master, but this might break so much stuff that I think it's best to add a constraint in stable for that.

Yoann-bary commented:
Okay, couple new things:
- Added a test for the constraints
- Removed part of a test that runs into the constraint which was added in the meantime
- Added the message to the .pot file (It's my first time adding a term so I'm unsure how to check if it worked, but it looks like other UserError in the code to me)

Bertrand2 commented:
> Added the message to the .pot file (It's my first time adding a term so I'm unsure how to check if it worked, but it looks like other UserError in the code to me)

Hello @Yoann-bary , how did you do it? If you never did it before, the proper way of adapting i18n is as follow:
- Run your server locally
- Go in debug mode
- Go in settings > translation > export translation
- Add the app you want to export (in this case "Time Off")
- Export the file
- Copy the content of the exported file in the pot file, but remove the changes of dates/version at the start of the file

If you did it like that, that should be all good! Otherwise I recommend to use that process to avoid any mistakes :)

Bertrand2 commented on addons/hr_holidays/tests/test_leave_requests.py:
```diff
@@ -1146,3 +1117,19 @@ def test_duration_display_global_leave(self):
+        # Does not raise an error since no leave of this type exists yet
+        self.holidays_type_1.requires_allocation = 'yes'
+
+        self.env['hr.leave'].create({
+            'name': 'Test leave',
+            'employee_id': self.employee_emp_id,
+            'holiday_status_id': self.holidays_type_1.id,
+            'date_from': (datetime.today() - relativedelta(days=1)),
+            'date_to': datetime.today(),
+            'number_of_days': 1,
+        })
```
this shouldn't work as there is no allocation for that leave type :thinking:

Yoann-bary replied:
It looks to me like there is no actual check for allocations here which is interesting. You are functionally prevented from selecting a leave type for which you have no allocation so you couldn't even make that request from the client side in the first place.

Still, this is pretty weird and I agree that I shouldn't have done that in my test. I'll change it

Bertrand2 replied:
I don't know how it was working exactly in 16.0 though I thought there was some check :thinking: , but either way you would have gotten an error raised in the forwardports since I'm certain there is a check at that point

Yoann-bary replied:
:+1:

Yoann-bary commented:
I'll do that and tell you once I'm done, Thanks

Yoann-bary commented:
> I'll do that and tell you once I'm done, Thanks

Should be good now

Bertrand2 reviewed (approved):
LGTM! Thanks for the changes :)

Bertrand2 commented:
robodoo r+
